Refactor: Hide `Gst.Sample` in `_Display` for simplicity

wmanley opened the pull request:
We were previously passing around `Gst.Sample`s because:

* There are timestamp metadata attached to them
* It was easier to handle lifetime management of the buffers, specifically WRT calling `gst_buffer_unmap`

These commits create a new numpy array type and a new backing store to solve each of those problems respectively.

Now this is done:

* We now no longer need to put `numpy_from_sample` everywhere.
* It opens the door to reducing copying, in the future we will be able to give clients these frames directly
* We have a place to attach arbitrary extra metadata to frames

I've implemented this for the last point.  I'll be needing that for #364.

TODO:

- [x] Release note: internal change

wmanley commented:
I've left a bunch of copies in that could potentially be removed, such as in `DeviceUnderTest.get_frame()`, `DeviceUnderTest.frames()` and `DeviceUnderTest._match_all()`.

wmanley commented:
@drothlis: Please review

drothlis commented on _stbt/gst_utils.py:
```diff
@@ -168,8 +168,27 @@ def __del__(self):
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+        self._gst_pts = getattr(obj, '_gst_pts', None)  # pylint: disable=protected-access,attribute-defined-outside-init
```
Was there documentation you followed on how to subclass an ndarray? Would be useful to mention that link in the commit message.

wmanley replied:
Yep: http://docs.scipy.org/doc/numpy-1.10.1/user/basics.subclassing.html#slightly-more-realistic-example-attribute-added-to-existing-array I'll add it when I rebase.

drothlis commented on _stbt/gst_utils.py:
```diff
@@ -168,8 +168,27 @@ def __del__(self):
 def array_from_sample(sample, readwrite=False):
-    return numpy.array(_MappedSample(sample, readwrite), copy=False)
+    f = numpy.array(
+        _MappedSample(sample, readwrite), copy=False).view(CapturedFrame)
```
Any particular reason why you do `numpy.array(args).view(CapturedFrame)` instead of `CapturedFrame(args)`?

wmanley replied:
I'll try it.

wmanley replied:
It works and looks much nicer.  Thanks

drothlis commented on _stbt/core.py:
```diff
@@ -738,7 +738,7 @@ def _match_all(self, image, frame, match_parameters, region):
             # sample when the contextmanager exits, but when no further
             # references are held to the numpy image.
-            npframe = numpy.copy(npframe)
+            npframe = npframe.copy()
```
This comment has now been implemented, but we still haven't removed the copy -- we should modify the comment to state when it will be safe to remove the copy.

drothlis replied:
Hmmm...  or has it. Why are we still using `numpy_from_sample`? (Here and in `Display.push_sample`.)

drothlis commented on _stbt/gst_utils.py:
```diff
@@ -168,8 +168,27 @@ def __del__(self):
             self._ctx.__exit__(None, None, None)
 
 
+class CapturedFrame(numpy.ndarray):
```
Presumably we're going to make this a public API at some future point (#364 perhaps).

wmanley replied:
Yes

drothlis commented on _stbt/core.py:
```diff
@@ -2619,17 +2616,16 @@ def test_wait_for_motion_half_motion_int():
-                    Gst.Sample.new(buf, Gst.Caps.from_string(
-                        'video/x-raw,format=BGR,width=2,height=2'), None, None))
+                yield CapturedFrame(
+                    data[(i // 2) % 2].flatten().reshape((2, 2, 3)),
```
Why is this flatten + reshape necessary, if `data[0]` and `data[1]` are already (2,2,3) numpy arrays?

wmanley replied:
Yep, nicer

drothlis commented:
Why do we still use `numpy_from_sample` in `Display.push_sample`? (And in `stbt_camera_calibrate.py` and `stbt_camera_validate.py`).

wmanley commented:
> Why do we still use numpy_from_sample in Display.push_sample? (And in stbt_camera_calibrate.py and stbt_camera_validate.py).

In these instances we're actually dealing with `Gst.Sample`s directly.  I guess they could be refactored too, I just didn't want to fiddle with code that needs to be `Gst.Sample` aware anyway.  I'll have a quick play.

wmanley commented:
> Why do we still use numpy_from_sample in Display.push_sample? (And in stbt_camera_calibrate.py and stbt_camera_validate.py).

Ok, I've replaced these too and removed `numpy_from_sample`.

wmanley commented:
Squashed.  @drothlis: You might like to look at the last commit before merging.

drothlis commented:
Looks good to me. Feel free to merge once travis has run.

wmanley commented:
Merged.  Some notes on possible future work:

- Replace `_AnnotatedTemplate` with an `ndarray` subclass
- Remove the copying when giving frames to users (grep for `.copy(`).  This will require validation that `push_sample` isn't doodling on frames that have already been passed to users.
